### ISSUE TITLE
Principal claim in 1.0.x

### ISSUE THREAD

**Issue by adrogon** [auth0/auth0-spring-security-api#24]
In 0.3.x, `Auth0SecurityConfig` was using `Auth0AuthenticationProvider` as `authenticationProvider`, which was setting `Principal` to `Auth0UserDetails`, which then set its `username` property based on several claims from JWT, especially `email`.
Long story short, injecting `Principal` in any endpoint and invoking `principal.getName()` returned the connected user's email if it is was present in the JWT claims.

In 1.0.x, the injected `Principal` is now an `AuthenticationJsonWebToken`, whose `getPrincipal()` function only returns the JWT `sub` claim.

What would you suggest is the easiest way to have an injected Principal rely on the JWT `email` claim (or another one) again?
- Set an `auth0.principalClaim` property in the app configuration that would select the intended claim to be used in `getPrincipal()`?
- Extend `AuthenticationJsonWebToken` and override its `getPrincipal()` method to look more like what `Auth0UserDetails` was doing with the `username` property in its constructor?
- Wait for you guys because that was already in the RC roadmap somehow :)

Spring Security ACL is relying on the Principal to retrieve the Sids, and we used to rely on it being set to the user's email. The current `sub` claim is not interesting enough to us.

Don't get me wrong, we are very glad to replace `Auth0SecurityConfig` by `WebSecurityConfigurerAdapter` :) I also can provide a PR based on your answer.

**Comment by arcseldon** [comment]
@adrogon - To clarify, and related to our conversation [here](https://github.com/auth0/auth0-spring-security-api/issues/23#issuecomment-273445271) - can you confirm you are sending a JWT Access Token, and using this API in conjunction with an Auth0 Resource Server?

The Access Token payload may look something like:

```
{
  "iss": "https://mytenant.auth0.com/",
  "sub": "auth0|585850xxxxxxxxx511f8",
  "aud": [
    "https://myresourceapi.com",
    "https://mytenant.auth0.com/userinfo"
  ],
  "azp": "mujye9zxsxxxxxxxxxxG15x9k",
  "exp": 1484796644,
  "iat": 1484760644,
  "scope": "openid profile read:profiles"
}
```

Above the user, logged into their SPA app (in this example) and then sent the received Access Token to this fictitious Resource API to get `profiles` information.  The endpoint was secured with `read:profiles` as a required scope. Based on this API, the request would be accepted so long as the access verification was approved (omitting other details for check this here). 

If your API Application somehow needs to lookup User Profile information - one option here might be to use the fact that the `Subject` claim is the UserID.  Then use the Management API from behind your trusted / secured API (right?) with a [get user by ID](https://auth0.com/docs/api/management/v2#!/Users/get_users_by_id) lookup. But that depends - do you actually need the user profile information, or were you searching for a way to avoid breaking changes to your existing application (refactor to a different approach). 

Note also, the above changes apply only to our API offering - the [Spring Security MVC](https://github.com/auth0/auth0-spring-security-mvc)  remains the same (for now) and still uses an approach quite similar to how this library used to work. The distinction being that the MVC Application is usually tied in some way to knowing the `authentication` details of the User. With a resource API, having knowledge of claims such as `email` may be of less interest, and rather it is the `authorization` that is important in determining which endpoints a request is able to consume.
Making sense?

**Comment by adrogon** [comment]
Hello @arcseldon,

I'm not sure I would call it the Auth0 Resource Server flow.
The link you provided [here](https://github.com/auth0-samples/auth0-spring-security-api-resource-server-sample) states the following:

> However, if you were just wishing to have something more akin to a headless Spring MVC application, or simple REST API, for example co-located with a Single Page Application (SPA) - then please also study this Auth0 Spring Security API Sample.

And that's what we went for.

Nonetheless, our current JWTs look like:
```
{
  "email": "my.email@gmail.com",
  "email_verified": true,
  "iss": "https://mytenant.eu.auth0.com/",
  "sub": "linkedin|s0m3th1nG",
  "aud": "API_KEY",
  "exp": 1484786242,
  "iat": 1484750242
}
```

In 0.3.x, `Principal.getName()` would return the value of the `email` claim, and now in 1.0.x, it contains the value of the `sub` claim.

We can still reach the `email` value by looking for that Claim in the DecodedJWT, but it would be easier if we could make `Principal` hold it like before.

The problem resides at [this line of code](https://github.com/auth0/auth0-spring-security-api/blob/master/lib/src/main/java/com/auth0/spring/security/api/authentication/AuthenticationJsonWebToken.java#L66) where `Principal` is set by default to the JWT's `subject`, and we were wondering if we should override that class so that it retrieves the `email` claim instead, or if that could be provided via a configuration property?

I hope this clarifies.

**Comment by arcseldon** [comment]
@adrogon - Thanks for clarifying. Again, as I stated when I shared the link - `using the legacy flows`.  The purpose of the new release - a Major semver update - is to move away from what we were previously encouraging users to do. 

Am assuming the token you pasted above is a JWT ID Token ?  In which case you are using the wrong library now - this library is intended to support the new flows which stipulate using a JWT Access Token (an ID Token should be rejected).

Importantly, also notice that the Access Token contains the Audience of the Resource Server

```
"aud": [
    "https://myresourceapi.com",
    "https://mytenant.auth0.com/userinfo"
  ]
```

You should continue to use the legacy API - which suited your needs - whilst you remain on legacy flows. If you are moving to the new pipeline unity (OIDC Conformant) flows - and I would greatly encourage you to do so - then you shall need to rethink your design accordingly. 

Thank you for the kind words with regards to version `0.3.x` - I crafted that library and pleased to hear it met your needs - it served its purpose for the flows we originally prescribed. Today we are heavily vested in making a clear understanding and distinction between OIDC and Oauth2 - and which JWT Tokens are appropriate for authentication vs authorization. 

In summary, SPA application should ALWAYS be using a Resource Server when making API requests, and not sending ID Tokens any longer under the new flows. You will find increasingly, as our samples, quickstarts and documentation catches up, that all mention of the old approach of sending ID Tokens will disappear. Support will remain for the old flows of course but they are considered deprecated and our literature / advise has already shifted.

Hope this all helps you understand the current state of play.

**Comment by adrogon** [comment]
Yes, I was expecting that answer, so I guess we'll begin to move on to stronger workflows if possible.

Thank you for your help.

**Comment by arcseldon** [comment]
@drogon - please take a look at this simple sample for illustration on how to use this new library with Access Token. Not sure it tells you anything you did not already know - but we were lacking a clear sample to show users. 

Procured this morning in response to your enquiries!

[Simple Sample](https://github.com/auth0-samples/auth0-spring-security-api-sample/tree/v1)

And thanks for your understanding with our evolution. Appreciate it can be painful when there are major semver updates etc, and you are already vested in a particular version.  Feel free to ping us if you need assistance with your migration (should you decide to go down that road soon).

**Comment by adrogon** [comment]
Let's say I believed 1.0 was more "ok, 0.x is stable enough to be version 1" than the semver breaking changing upgrade ;) 

Thank you for the example. I understand what you are achieving here, we simply didn't know about [this](https://auth0.com/docs/api-auth/tutorials/configuring-tenant-for-api-auth).
We'll see if we can build our architecture on top of it.
